platform/src/main/java/org/hillview/storage/CassandraSSTableLoader.java

```diff
@@ -171,137 +232,236 @@ public Schema getSchema(CFMetaData metadata) {
-    private List<IAppendableColumn> loadColumns(boolean[] columnToLoad, Stream<UnfilteredRowIterator> partitions, int size) {
-        List<IAppendableColumn> columns = createColumns(columnToLoad, size);
-        // Iterating each item inside the table
-        partitions.forEach(partition -> {
-            while (partition.hasNext()) {
-                Unfiltered unfiltered = partition.next();
-                if (unfiltered instanceof Row) {
-                    Row row = (Row) unfiltered;
-                    int i = 0;
-                    int currentColumn = 0;
-                    Object value;
-                    IAppendableColumn col;
-                    // Extracting the value for each column from a single row
-                    for (ColumnData cd : row) {
-                        // Filter the column to only load the one marked true by columnToLoad[]
-                        if (columnToLoad[i]) {
-                            if (cd.column().isSimple()) {
-                                Cell cell = (Cell) cd;
-                                AbstractType<?> cellType = cell.column().cellValueType();
-                                value = cellType.getSerializer().deserialize(cell.value());
+    /**
+     * This function will load specific column listed at columnDefinitions and in certain token ranges 
+     */
+    private static void loadColumns(SSTableReader ssTableReader, List<IAppendableColumn> columns,
+            PartitionColumns columnDefinitions, List<CassandraTokenRange> tokenRanges, String localEndpoint) {
+        ColumnFilter cf = ColumnFilter.selection(columnDefinitions);
+        TypeSerializer<?>[] arrSerializers = new TypeSerializer<?>[columnDefinitions.size()];
+        CQL3Type[] arrColTypes = new CQL3Type[columnDefinitions.size()];
+        Iterator<ColumnDefinition> colIter = columnDefinitions.regulars.simpleColumns();
+        int i = 0;
+        // init reusable serializer and prefixComparison
+        while (colIter.hasNext()) {
+            ColumnDefinition cd = colIter.next();
+            CQL3Type colType = cd.type.asCQL3Type();
+            arrColTypes[i] = colType;
+            arrSerializers[i] = colType.getType().getSerializer();
+            i++;
+        }
+
+        for (CassandraTokenRange tr : tokenRanges) {
+            // Load the partition if the local endpoint is listed as the first endpoint
+            if (!tr.endpoints.get(0).equals(localEndpoint)) continue;
+            DataRange range = DataRange.forTokenRange(tr.tokenRange);
+            SSTableReadsListener listener = CassandraSSTableLoader.newReadCountUpdater();
+            ISSTableScanner currentScanner = ssTableReader.getScanner(cf, range, false, listener);
+            Spliterators.spliteratorUnknownSize(currentScanner, Spliterator.CONCURRENT)
+                .forEachRemaining((partition) -> {
+                    Unfiltered unfiltered = partition.next();
+                    if (unfiltered instanceof Row) {
```

Comment:
Could you assert here that unfiltered is an instance of Row?
What else could it be?

Reply:
The other possibility is to have `RangeTombstoneMarker`. This is the original Cassandra's code to read the sstable:
```
Unfiltered unfiltered;
while (partition.hasNext())
{
    unfiltered = partition.next();
    if (unfiltered instanceof Row)
    {
        serializeRow((Row) unfiltered);
    }
    else if (unfiltered instanceof RangeTombstoneMarker)
    {
        serializeTombstone((RangeTombstoneMarker) unfiltered);
    }
    updatePosition();
}
```